---
title: "Cm_Pac_msat_EDA"
author: "Lisa Komoroske"
date: "Nov 14, 2017"
output: html_document
---

Intro: This script is for QA/QC and basic population structure analyses using the package StrataG (see E. Archer's github for details on this package)

This is meant to be a template/example that can be amended for other datasets-(currently code matches a Cm turtle microsatellite dataset)  

### Setup: Load packages etc
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# make sure you have Rtools installed
#if (!require('devtools')) install.packages('devtools')#can comment out if/once you have installed
# install from GitHub
#devtools::install_github('ericarcher/strataG', build_vignettes = TRUE)#need internet connection. don't need to run
#every time, just if you think Eric's made changes and you want to update
library(strataG)
library(plyr)
rm(list = ls())
options(stringsAsFactors = F)
#setwd('~/...')#change this accordingly
getwd()#this just double checks that you are where you wanted to be
```

### Intro: set keys and load data
```{r intro, include=TRUE}

description <- "Pac_CmMsat_2017-11-14_EDA"
strata.file <- "Pac_Cm_Strata_111317.csv"
genotypes.file <- "Pac_Cm_genodata_111317.csv"
geno.data.start<-3 #make this number match the last column before genotype data
#read in data ####
Msat.Cm.geno <- readGenData(genotypes.file) #analysis set export with population column removed
#Msat.Cm.geno<-subset(Msat.Cm.geno,Dups=="N")#remove previously identified/marked duplicates
    #missing genotype must be "NA" (not "0")
Msat.Cm.strata <- readGenData(strata.file) #LABID and population from analysis set.
#Msat.Cm.strata <- subset(Msat.Cm.strata,Dups=="N")

strata.schemes <- Msat.Cm.strata[, c("Broad","Fine")]#note-here you can have additional columns, and then this makes it easy to switch between strata levels below
rownames(strata.schemes) <- Msat.Cm.strata$Lab_ID

head(Msat.Cm.geno)
#change 0s to NAs (if needed)
Msat.Cm.geno[Msat.Cm.geno==0]=NA
#check that they are NAs
View(Msat.Cm.geno)
```

### Data QC: Check for samples missing a lot of data  
This takes out cases that are missing more data for than X% of the markers. You can change the percent in the formula as you like.  
***(Note that we are counting the NAs, so the percent is the inversion, i.e. if more than 30% of the loci have NAs-we have data for less than 70% of the markers)***
```{r dataQC1, include=TRUE}
#Run these lines if you want to take out the samples that have missing data over a certain percent: #####
numNAs<-apply(Msat.Cm.geno, 1, function(z) sum(is.na(z)))
Msat.Cm.geno.a<-Msat.Cm.geno[!(numNAs>((ncol(Msat.Cm.geno)-geno.data.start))*0.3),]#Sue, I reset this to 0.3 because that is equivalent to 6 missing alleles (i.e., max you can be missing is 3 markers-I looked quickly didn't see any cases where only one allele was scored for a marker...if this is not the case you can adjust the criteria here)

```

### Data QC: check for bad strata with low numbers of samples  
This shows you the counts for each location, can then subset out  
```{r dataQC2, include=TRUE}
dir.create(description)#makes a directory to put all your output files for organization purposes
setwd(paste(getwd(),"/",description,sep=""))#change your WD so all files get put there from this run
#note this just a quick look by strata-doesn't account for any samples that were dropped from above due to a lot of missing data
t1<-table(Msat.Cm.geno.a$Fine) 
write.csv(t1, file = paste(description, "_counts_bystrata_afterNAremoval.csv", sep = ""))

#run if want to remove samples where have less than X in a strata or missing data for an entire locus:
#Msat.Cm.geno.short<-subset(Msat.Cm.geno.a,Broad!="XX")

```

### Create a diploid gtypes object####
```{r gtypes, include=TRUE}
Msat.g <- df2gtypes(Msat.Cm.geno.a, ploidy = 2, id.col=3, loc.col=4,schemes = strata.schemes)
Msat.g <- stratify(Msat.g, "Fine")
summary(Msat.g) #this gives a quick summary
save(Msat.g, file = paste(description, "_gtypes.rdata", sep=""))
```

**Note-amend to do for ones with/without small groups removed, etc.**

### Test for all QA/QC tests: locus summaries, check for duplicates, HWE, etc.  
```{r gtypesQAQC, include=TRUE}
qaqc(Msat.g, label = paste(description, "_QC_results", sep=""), num.shared = 0.9) 
#Note-in the locus summary files:
#1 the propUniqueAlleles is more useful for haplotype (mitochondrial) bc it is the percentage of alleles at that locus that only occur once (so for diploid markers, this=one individual is heterozygous)
#2 the allelic richness of each locus calculated as the number of alleles divided by the number of samples without missing data at that locus.

#check for private alleles:
pA<-privateAlleles(Msat.g)
write.csv(pA, file = paste(description, "_private_alleles.csv", sep = ""))

#check for LD:
LD<-LDgenepop(Msat.g)#for all together; but this is not really helpful since if there's underlying structure it can make them come out as linked; need to do by strata
write.csv(LD, file = paste(description, "_all_LD.csv", sep = ""))
#x <- strataSplit(Msat.g)#split up by strata, and then can call it to do something
#by strata-
#test first with something easy like just printing num Alleles
for(g in strataSplit(Msat.g)) print(numAlleles(g))#this is how you loop it through all the strata, use print if you just want to see it
#to write them to file:
for(g in strataSplit(Msat.g)) {
  LD <- LDgenepop(g)
  fname <- paste(description,"_LD_", strataNames(g)[1], ".csv", sep = "")
  write.csv(LD, fname)
}
# HWE-
#for all together-
HWE<-hweTest(Msat.g, use.genepop = TRUE, label = "HWE.genepop")#note the default is genepop is false bc eric didnt want to assume people had that, but is actually more accurate so change to true
write.csv(HWE, file = paste(description, "_all_HWE.csv", sep = ""))

#looping through each strata...
for(g in strataSplit(Msat.g)) {
  hwe <- hweTest(g, use.genepop = T)
  na <- numAlleles(g)
  result <- cbind(num.alleles = na, hwe)
  fname <- paste(description,"_HWE_ ", strataNames(g)[1], ".csv", sep = "")
  write.csv(result, fname)
}# or use this if want to run and bind multiple things together that have the same dimensions

#Allele Frequencies:
AF<-alleleFreqs(Msat.g, by.strata = FALSE)# list of allele frequencies for each locus. Each element is a matrix or array with frequencies by count (freq) and proportion (prop) of each allele.
for(loc in names(AF)) {
  fname <- paste(description, loc, "overall_allele.freqs.csv", sep = "_")
  write.csv(AF[[loc]], file = fname)
}
#strata=TRUE if want by to do for each strata
af <- alleleFreqs(Msat.g, by.strata=T)
for(loc in names(af)) {
  fname <- paste(description, loc, "bystrata_allele.freqs.csv", sep = "_")
  write.csv(af[[loc]][, "freq", ], file = fname)
}

#this jackknifes n samples at a time (set to 1) to see how affects HWE 
Msat.Cm.JackHWE<-jackHWE(Msat.g, exclude.num = 1, min.hwe.samples = 3, show.progress = TRUE,
        use.genepop = TRUE)#performs a HWE jackknife where all combinations of exclude.num samples are left out and HWE is recalculated
Msat.Cm.JackHWE.influential<-jackInfluential(Msat.Cm.JackHWE, alpha = 0.05)
Msat.Cm.JackHWE.influential #HWE and identifies "influential" samples. Samples are "influential" if the observed HWE p-value is < alpha, but is > alpha when the samples are not present.
write.csv(Msat.Cm.JackHWE.influential$influential, file = paste(description, "_JackHWE.influential.csv", sep = ""))
write.csv(Msat.Cm.JackHWE.influential$allele.freqs, file = paste(description, "_JackHWE.allele.freqs.csv", sep = ""))
write.csv(Msat.Cm.JackHWE.influential$odds.ratio, file = paste(description, "_JackHWE.odds.ratio.csv", sep = ""))

pdf(paste(description, "_JackHWE.influential", ".pdf", sep = ""), width = 15, height = 8)
#creates a cumulative frequency plot of all odds-ratios from jack.influential. 
#A vertical dashed line marks the smallest influential exclusion
plot(Msat.Cm.JackHWE.influential, main = "Msat.Cm.JackHWE.influential")
dev.off()
```

# Run just overall, pairwise and/or both tests for all or specified metrics####
```{r popstructure, include=TRUE}
#All:
#popStructTest(Msat.Cm.g.Pop, nrep = 10, stats = "all", type = c("both", "overall","pairwise"), keep.null = FALSE, quietly = FALSE, num.cores = 1,write.output = FALSE)

#Overall:
overall.1 <- overallTest(Msat.g, stats = "all", nrep = 10000, write.output=FALSE)
write.csv(overall.1$result, file = paste(description, "_overall_test.csv", sep = ""))
#test with nrep <100; use nrep=1000 or 10000
#write.output=TRUE prints pairwise matrix for each test, with test value in the upper right
#quadrant and p-values in the lower left quadrant (only makes sense for pairwise)

# Run just pairwise tests for specified metrics####
pairwise.1 <- pairwiseTest(Msat.g, stats = c("fst", "fst.prime","gst.dbl.prime"),
                            nrep = 10000, write.output=TRUE) #set nrep = 10000 after testing at <100
write.csv(pairwise.1$result, file = paste(description, "_pairwise_results.csv", sep = ""))
warnings()
#Can save to file Warning messages about dropped loci to remember if applicable

#Calculate observed heterozygosity for each locus ()####
oh <- obsvdHet(Msat.g)
write.csv(oh, paste(description, "_Ho.csv", sep = ""))

#oh here is the same as that found in the qaqc output files, this just makes it available for easy graphing manipulation
```

### Follow up scripts:
1. See 'Genetic_graphs_EA_LK_incorporated.rmd' in Pop genomics repository -Genetics graphs- integrating StrataG and ggplots to generate graphics of interest for basic genetic analyses:add in as needed here.
2. STRUCTURE, PHASE scripts-to run these other programs through StrataG:add in as needed here.